crates/marine-test-macro-impl/src/marine_test/token_stream_generator/service_generator.rs

```diff
@@ -0,0 +1,282 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::attributes::{ServiceDescription};
+use crate::TResult;
+use crate::TestGeneratorError;
+use crate::marine_test::config_utils::{Module, ProcessedService};
+use crate::marine_test::utils::new_ident;
+use crate::marine_test::{modules_linker, config_utils};
+
+use proc_macro2::TokenStream;
+use quote::quote;
+
+use std::path::Path;
+use std::path::PathBuf;
+use itertools::{Itertools, zip};
+use std::collections::HashMap;
+use crate::marine_test::modules_linker::{LinkedModules, LinkedModule, UseDescription};
+use marine_it_parser::it_interface::IModuleInterface;
+
+pub(crate) fn generate_service_definitions(
+    services: HashMap<String, ServiceDescription>,
+    file_path: &PathBuf,
+) -> TResult<Vec<TokenStream>> {
+    let services = services
+        .into_iter()
+        .sorted_by(|lhs, rhs| lhs.0.cmp(&rhs.0))
+        .map(|(name, service)| ProcessedService::new(service, name, file_path))
+        .collect::<TResult<Vec<ProcessedService>>>()?;
+
+    let service_modules = services
+        .iter()
+        .map(|service| {
+            let modules_dir_test_relative = file_path.join(&service.config.resolved_modules_dir);
+            let modules =
+                config_utils::collect_modules(&service.config.config, &modules_dir_test_relative)?;
+            Ok(modules)
+        })
+        .collect::<TResult<Vec<Vec<Module<'_>>>>>()?;
+
+    let link_info = link_services(zip(&services, &service_modules))?;
+    services
+        .iter()
+        .map(|service| -> TResult<TokenStream> {
+            let service_mod = new_ident(&service.name)?;
+            // entry with service.name was added in link_services(...), so unwrap is safe
+            let service_definition = generate_service_definition(
+                &service,
+                file_path,
+                link_info.get::<str>(&service.name).unwrap(),
+            )?;
+            let glue_code = quote! {
+                pub mod #service_mod {
+                    #service_definition
+                }
+            };
+            Ok(glue_code)
+        })
+        .collect::<TResult<Vec<TokenStream>>>()
+}
+
+fn link_services<'modules>(
+    services: impl ExactSizeIterator<
+        Item = (&'modules ProcessedService, &'modules Vec<Module<'modules>>),
+    >,
+) -> TResult<LinkedModules<'modules>> {
+    let facade_modules = services
+        .map(|(service, modules)| {
+            let facade = match modules.last() {
```

Comment:
let's add  a comment that facade module is a last one, or even write a function with such a name